### PR TITLE
SB-100 setting up unit tests heartbeat

### PR DESCRIPTION
**Pull request by StudentCZ**
Got 2 working mock GET unit test and 2 mock POST unit test. Testing to make sure the function in the app.py still works. Testing different edge cases. 

Screen Shot 2023-01-16 at 4 48 43 PM

Here is a snippet of a few unit tests
Screen Shot 2023-01-16 at 4 57 58 PM

### CODE REVIEW

**Review by FioraDeMarco — APPROVED** [review]

File: test_unit.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        data = {
+            "user_id": "123",
+            "time_stamp": 1669980264,
+            "latitude": "37.7749",
+            "longitude": "-122.4194",
+            "speed": 0
+        }
+        response = client.post("/heartbeat", json=data)
+        assert response.status_code == 200
+        assert response.data == b'Heartbeat data added successfully'
```
[the remark this answers is not in the document]
Comment (StudentCZ) [reply test_unit.py:47]: You are asking about integration tests. Integration tests actually interactive with the DB to see if the DB actually have our data. This just mock our function in app.py and assume that everything is working. Below is a screenshot of a mock post unit test to see how our data should look when we do a mock post into the DB. 

If we want to implement integration tests and make an actual post to the DB we will have to change and add more to the unit test code.


If I change user_id to userId we get this error. 
Screen Shot 2023-01-17 at 12 22 57 AM


But if we input the key/values pair as intended we get a passing test.
Screen Shot 2023-01-17 at 12 24 03 AM

File: test_unit.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        get_latest_heartbeat_mock.assert_called_once()
+        assert response.status_code == 200
+        assert response.get_json() == expected_response
+
+def test_get_heartbeats_by_user_id_not_found(client):
+    with patch("app.get_latest_heartbeats") as get_latest_heartbeat_mock:
+        get_latest_heartbeat_mock.return_value = {"Items":[]}
+        response = client.get("/heartbeat/123")
+        get_latest_heartbeat_mock.assert_called_once()
+        assert response.status_code == 200
```
Comment (henryluan95) [inline test_unit.py:33]: Should the status code for this test be 404 (Resource not found)?

Comment (chunloy) [reply test_unit.py:33]: It seems to me that it will just return empty data which I think is still valid?

Comment (henryluan95) [reply test_unit.py:33]: Hmmm make sense. 200 can indicate that the table exists but the requested data isn't there.

File: test_unit.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    "user_id": "",
+    "time_stamp": "",
+    "latitude": "",
+    "longitude": "",
+    "speed": ""
+  }
+
+  response= client.post("/heartbeat", json=data)
+  assert response.status_code == 400
+  assert response.data == b"Unable to write"
```
Comment (chunloy) [inline test_unit.py:60]: Didn't catch this before. Could we make this response more descriptive e.g. "Unable to write data to heartbeat table."

**Review by chunloy — DISMISSED** [review]
Great job overall. Left a comment in the unit test file about updating the error message to be more descriptive.

**Review by yhus — APPROVED** [review]
LGTM!

File: test_unit.py (hunk 3 above)
Comment (StudentCZ) [reply test_unit.py]: Updated the message.
Screen Shot 2023-02-02 at 11 24 39 PM
